transactions/src/main/java/org/jboss/as/txn/service/internal/tsr/JCAOrderedLastSynchronizationList.java

```diff
@@ -87,7 +107,18 @@ public void addInterposedSynchronization(Synchronization synchronization) {
      */
     @Override
     public void beforeCompletion() {
-        for (Synchronization preJcaSync : preJcaSyncs) {
+        // This is needed to guard against syncs being registered during the run
+        int lastIndexProcessed = -1;
+        Synchronization[] copiedPreJcaSyncs = (Synchronization[]) preJcaSyncs.toArray(new Synchronization[] {});
```

Comment:
I was working from the way it is implemented in Narayana today (https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java#L362). However, I have revised my patch to just use the index.